Case study on nextjs server and client side rendering (#309)
NoroffFEU/development-platforms: src/content/docs/frameworks/Next.js/NextJs-rendering-case-study.md

---
title: Next.JS server-side vs client-side rendering
tags: nextJS, frameworks, SEO, rendering, client, server.
---

## Introduction

One of the core decisions web developers must make is where to implement logic and rendering in their application. This can be difficult because there are so many ways to build a website.

> Rendering converts the code you write into user interfaces. React and Next.js allow you to create hybrid web applications where parts of your code can be rendered on the server or the client. In Next.JS, you can choose where you want your components to be rendered: on the server side or on the client side. This study will help you understand the differences between these rendering environments, strategies, and runtimes.

Reviewer: ```suggestion
Rendering converts the code you write into user interfaces. React and Next.js allow you to create hybrid web applications where parts of your code can be rendered on the server or the client. In Next.js, you can choose where you want your components to be rendered: on the server side or on the client side. This study will help you understand the differences between these rendering environments, strategies, and runtimes.
```

### Advantages and disadvantages of using client-side components

**Pros:**

**Interactivity:** Client Components can use state, effects, and event listeners, meaning they can provide immediate feedback to the user and update the UI.
**Browser APIs:** Client Components have access to browser APIs, like geolocation or localStorage.

**Cons:**

The primary downside to client-side rendering is that the amount of JavaScript required tends to grow as an application grows, which can impact a page's INP ('Interaction to Next Paint'). This becomes especially difficult with the addition of new JavaScript libraries, polyfills, and third-party code, which compete for processing power and must often be processed before a page's content can render. Another disadvantage are poor Search Engine Optimization (SEO) , as crawlers have hard time crawling client rendered content.

> _How to use client-side components in Next.JS_

Reviewer: ```suggestion
_How to use client-side components in Next.js_
```

and so on could be rendered on the server side. This limitation led to challenges such as prop drilling and duplication of
API calls when passing data to lower-level components.

### Advantages and disadvantages of using server-side components

**Pros:**

**Efficient utilisation of server resources:** Fetch data and access all backend resources while rendering in server with less latency, and can be directly displayed to client reducing loading time.
**SEO Friendliness:** Search engines favor pages with well-structured HTML content. Since SSR provides pre-rendered HTML, it enhances the SEO of Next.js applications, making them more discoverable by search engines.
**Maintainability and Consistency:** With SSR, you can centralize data fetching and processing logic on the server, promoting a cleaner and more maintainable codebase. This approach ensures consistency between server-rendered content and client-rendered content.
**Security:** Safe to use API key within server components.

**Cons**

When it comes to interactivity, server components can be slower and less interactive than client components. You also can only use React hooks from a client component.

> _How to use server-side components in Next.JS_

Reviewer: ```suggestion
_How to use server-side components in Next.js_
```

By default, Next.js uses Server Components. This allows you to automatically implement server rendering with no additional configuration, and you can opt into using Client Components when needed.

## When To Use Which

> It can be confusing on when to use which, so there is a chart from official nextJS docs:

Reviewer: ```suggestion
It can be confusing on when to use which, so there is a chart from official Next.js docs:
```

- The output of the server-rendered components.
- Placeholders indicating where the client components should go, along with references to their JavaScript files.

React then uses this data to combine the server and client components and update the webpage in the browser.

One misconception is that 'use client' is Client Side only. That's not really the case. 'use client' is a hint to the the bundler, that the module has to be part of the JavaScript that is sent to the client. Server Components, never share their JavaScript with the client.

## Conclusion

use-client and use-server aren't just directives; they are powerful tools for Next.js developers. By understanding their individual strengths and using them strategically, you can create applications that excel in performance, SEO, and user experience. Remember, the key lies in finding the right balance and tapping into the full range of rendering possibilities that Next.js 14 offers.

> From use-client & use-server in next Js 14 to mastering the rendering game with Next.js 14, What’s New, What’s the Difference, and Why Does It Matter? is an exciting journey, full of twists and turns that unlock incredible possibilities.

Reviewer: This is a confusing sentence.
